torch/csrc/jit/export.cpp

```diff
@@ -634,6 +636,95 @@ void ScriptModuleSerializer::writeTensorTable(torch::ModelDef* model_def) {
   }
 }
 
+size_t pushBytes(size_t num_bytes, std::vector<char>& data) {
+  AT_ASSERT(num_bytes > 0);
+  data.push_back(0);
+  auto index = data.size() - 1;
+
+  for (size_t i = 1; i < num_bytes; ++i) {
+    data.push_back(0);
+  }
+  return index;
+}
+
+template<typename T>
+size_t pushValue(T value, std::vector<char>& data) {
+  auto index = pushBytes(sizeof(T), data);
+  T* ptr = reinterpret_cast<T*>(&data[index]);
+  *ptr = value;
+  return index;
+}
+
+// Estimate the number of bytes a serialized IValue will take
+size_t estimateIValueSize(const IValue& ivalue) {
+  if (ivalue.isString()) {
+    return ivalue.toStringRef().size() + 1;
+  } else if (ivalue.isGenericDict()) {
+    auto dict = ivalue.toGenericDictRef();
+    size_t dict_overhead = 2 * sizeof(uint64_t);
+    size_t element_size = 0;
+    if (dict.begin() != dict.end()) {
+      const auto item = dict.begin();
+      element_size =
+          estimateIValueSize(item->first) + estimateIValueSize(item->second);
+    }
+    return dict_overhead + element_size * dict.size();
+  }
+  return 0;
+}
+
+uint64_t ScriptModuleSerializer::writeIValue(
```

Comment:
A string (or any IValue) that appears twice that is going to be serialized twice in this design. Non-primitive objects (anything that cannot fit into a single IValue) should be written once per zip file.

Reply:
oh, you want to recover original sharing of string instances? It's probably the case for dicts indeed